src/library/Cardano/CoinSelection.hs

```diff
@@ -226,51 +271,98 @@ sumOutputs =  coinMapValue . outputs
 sumChange :: CoinSelection i o -> Coin
 sumChange = mconcat . change
 
--- | Represents a set of options to be passed to a coin selection algorithm.
+-- | Defines an __inclusive upper bound__ on the /number/ of inputs that
+--   a 'CoinSelectionAlgorithm' is allowed to select.
 --
-data CoinSelectionOptions i o e = CoinSelectionOptions
-    { maximumInputCount
+newtype CoinSelectionLimit = CoinSelectionLimit
+    { calculateLimit
         :: Word8 -> Word8
             -- ^ Calculate the maximum number of inputs allowed for a given
             -- number of outputs.
-    , validate
-        :: CoinSelection i o -> Either e ()
-            -- ^ Validate the given coin selection, returning a backend-specific
-            -- error.
-    } deriving (Generic)
+    } deriving Generic
 
 -- | Represents the set of possible failures that can occur when attempting
---   to produce a 'CoinSelection'.
---
-data CoinSelectionError e
-    = ErrUtxoBalanceInsufficient Coin Coin
-    -- ^ The UTxO balance was insufficient to cover the total payment amount.
-    --
-    -- Records the /UTxO balance/, as well as the /total value/ of the payment
-    -- we tried to make.
-    --
-    | ErrUtxoNotFragmentedEnough Natural Natural
-    -- ^ The UTxO was not fragmented enough to support the required number of
-    -- transaction outputs.
-    --
-    -- Records the /number/ of UTxO entries, as well as the /number/ of the
-    -- transaction outputs.
-    --
-    | ErrUtxoFullyDepleted
-    -- ^ Due to the particular distribution of values within the UTxO set, all
-    -- available UTxO entries were depleted before all the requested
-    -- transaction outputs could be paid for.
-    --
-    | ErrMaximumInputCountExceeded Natural
-    -- ^ The number of UTxO entries needed to cover the requested payment
-    -- exceeded the upper limit specified by 'maximumInputCount'.
-    --
-    -- Records the value of 'maximumInputCount'.
-    --
-    | ErrInvalidSelection e
-    -- ^ The coin selection generated was reported to be invalid by the backend.
-    --
-    -- Records the /backend-specific error/ that occurred while attempting to
-    -- validate the selection.
-    --
-    deriving (Show, Eq)
+--   to produce a 'CoinSelection' with a 'CoinSelectionAlgorithm'.
+--
+-- See 'selectCoins'.
+--
+data CoinSelectionError
+    = InputValueInsufficient
+        InputValueInsufficientError
+    | InputCountInsufficient
+        InputCountInsufficientError
+    | InputLimitExceeded
+        InputLimitExceededError
+    | InputsExhausted
+        InputsExhaustedError
```

Comment:
This does seem to make the Haddocks easier to read.

Reply:
Yes. It seems Haddock doesn't cope well with paragraphs within error constructors. (As the text is squashed within a table cell.)

Another benefit of giving each error a separate type is that we can also give names to error parameters, which should help with debugging.